### ISSUE TITLE
Remove @Deprecated on GraphQLScalarType constructor

### ISSUE THREAD

**Issue by etienne-sf** [graphql-java/graphql-java#1777]
Hello,

You're apparently in the process to move from the use of constructors to the use of Builder, to instanciate GraphQLScalarType.

I'm working on the https://github.com/graphql-java-generator/graphql-maven-plugin-project project. This project is based on graphql-java, and uses graphql-java-spring-boot-starter-webmvc when in server mode.

The aim is to separate what's linked to GraphQL and what's provided by the application.

The GraphQL configuration respects graphql-java-spring-boot-starter-webmvc, and is provided as Spring Beans : that's nice.
The application configuration SHOULD be provided as spring Component. This includes Custom Scalars implemented by the application.

This last point becomes impossible, without the GraphQLScalarType constructors.

So please, keep GraphQLScalarType constructors (remove the deprecation), and allow the application to choose between Spring Bean and Component.

Etienne

**Comment by andimarek** [comment]
@etienne-sf can you explain a bit more why you need a constructor and can't use the Builder? thanks

**Comment by etienne-sf** [comment]
The aim of Spring is to have the minimum link between components.

So just declaring a class as beeing a @Component (of type GraphQLScalarType, here) should be enough. Then the framework would gather all component of this type, and wire them as scalars.

But when using this builder, you to code the wiring. 

So I can manage the Builder.


But I find useless to have all developpers to manually wire the scalars.
Just creating a subclass of GraphQLScalarType, and mark it as a @Component should be enough.

For that, the constructors of GraphQLScalarType should be public.


What's your opinion on this ?

**Comment by bbakerman** [comment]
Why is this bad?

```
    @Bean(name = "CustomScalar")
    public GraphQLScalarType myCustomScalar() {
        Coercing coercing = new Coercing() {
            @Override
            public Object serialize(Object dataFetcherResult) throws CoercingSerializeException {
                return "something";
            }

            @Override
            public Object parseValue(Object input) throws CoercingParseValueException {
                return "acceptSomething";
            }

            @Override
            public Object parseLiteral(Object input) throws CoercingParseLiteralException {
                return "astSomething";
            }
        };
        return GraphQLScalarType.newScalar().name("custom")
                .coercing(coercing)
                .description("Some decription about it")
                .build();
    } 
    
    {
        ApplicationContext appCtx = ....;

        Map<String, GraphQLScalarType> beansOfType = appCtx.getBeansOfType(GraphQLScalarType.class);
    }

```

Spring can get things without having to have a named class nor a directly instantiated object.

**Comment by etienne-sf** [comment]
Yes.

But typically, @Bean is the replacement for the XML Spring configuration file.
All applicative component should be as little coupled as possible.

So having to call the scalar method of the runtime wiring is too complex.

Just discovering the @Component of type GraphQLScalarType is either, and make the code less coupled.

The runtime wiring (hard coded link between java classes) is replaced by a runtime disovery of Component.

It's lighter and easier to use.

So it's better...
:)

But to do so, you must be able to call the construction for GraphQLScalarType.

Etienne

**Comment by etienne-sf** [comment]
BTW could you just remove the deprecation on GraphQLScalarType, and let developper choose ?

Why forbid developers to create subclasses of GraphQLScalarType  ?

**Comment by lqc** [comment]
@etienne-sf Using builders in Spring configuration is not something new either. 

GraphQLScalarType is designed to be an immutable object and those typically don't allow subclassing, because you no longer can guarantee that a subclass is immutable or that it implementes equals() correctly.

**Comment by etienne-sf** [comment]
I think that this point can be solved by defining equals as final.

**Comment by bbakerman** [comment]
One thing we have learned in maintaining a library API is that a constructor is actually SPI.

We cannot add new members to it if people can call them directly.

The builder pattern shields you from this - so we can do whatever tricks we need to add new members without breaking consumers.

So we tend to (but not always) use builders and make the constructor private.  If we started again (knowing what we know now) virtually all of the graphql code base would use builders for construction.

Hence we deprecated to NOT break existing people.
